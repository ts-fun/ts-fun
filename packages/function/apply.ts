/**
 * Apply a function with an array of arguments
 * @param fn Function to be applied
 * @param xs Arguments to be passed to `fn`
 * @returns Result of function application
 */
export const apply = <Xs extends readonly any[], Y>(
  fn: (...xs: Xs) => Y,
  xs: Xs,
) => fn(...xs)

/**
 * Apply a function with an iterable of arguments
 * @param fn Function to be applied
 * @param xs Arguments to be passed to `fn`
 * @returns Result of function application
 */
export const applyIter: (
  <X, Y>(
    fn: (...xs: X[]) => Y,
    xs: Iterable<X>,
  ) => Y
) = apply as any
