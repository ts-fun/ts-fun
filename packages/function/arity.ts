/**
 * Swap arguments of a binary function
 * @param fn Function to swap arguments
 * @returns Function that calls `fn`
 */
export const flip =
  <X0, X1, Y> (fn: (x0: X0, x1: X1) => Y) =>
    (x1: X1, x0: X0) => fn(x0, x1)

/**
 * Swap the first 2 arguments of an n-ary function
 * @param fn Function to swap arguments
 * @returns Function that calls `fn`
 */
export const flipXs =
  <X0, X1, Xs extends any[], Y>
    (fn: (x0: X0, x1: X1, ...xs: Xs) => Y) =>
      (x1: X1, x0: X0, ...xs: Xs) => fn(x0, x1, ...xs)

/**
 * Turn a binary function into n-ary
 * by calling that function multiple times
 * @param fn Function to apply
 * @returns Function that calls `fn` multiple times
 */
export const nAry =
  <X> (fn: (x0: X, x1: X) => X) =>
    (...xs: [X, X, ...X[]]) => xs.reduce((x0, x1) => fn(x0, x1))

/**
 * Turn a binary function into n-ary
 * by calling that function multiple times
 * @param fn Function to apply
 * @returns Function that calls `fn` multiple times
 */
export const nAryRight =
  <X> (fn: (x0: X, x1: X) => X) =>
    (...xs: [X, X, ...X[]]) => xs.reduceRight((x0, x1) => fn(x0, x1))
