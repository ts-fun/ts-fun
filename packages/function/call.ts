/**
 * Call a nullary function
 * @param fn Function to call
 * @returns Result of function execution
 */
export const call0 = <Y> (fn: () => Y) => fn()

/**
 * Call a unary function
 * @param fn Function to call
 * @param x Argument to pass to `fn`
 * @returns Result of function execution
 */
export const call1 = <X, Y> (
  fn: (x: X) => Y,
  x: X
) => fn(x)

/**
 * Call a binary function
 * @param fn Function to call
 * @param x0 First argument to pass to `fn`
 * @param x1 Second argument to pass to `fn`
 * @returns Result of function execution
 */
export const call2 = <X0, X1, Y> (
  fn: (x0: X0, x1: X1) => Y,
  x0: X0,
  x1: X1
) => fn(x0, x1)

/**
 * Call a ternary function
 * @param fn Function to call
 * @param x0 First argument to pass to `fn`
 * @param x1 Second argument to pass to `fn`
 * @param x2 Third argument to pass to `fn`
 * @returns Result of function execution
 */
export const call3 = <X0, X1, X2, Y> (
  fn: (x0: X0, x1: X1, x2: X2) => Y,
  x0: X0,
  x1: X1,
  x2: X2
) => fn(x0, x1, x2)

/**
 * Call a n-ary function
 * @param fn Function to call
 * @param args Arguments to pass to `fn`
 * @returns Result of function execution
 */
export const callXs = <Xs extends any[], Y> (
  fn: (...args: Xs) => Y,
  ...args: Xs
) => fn(...args)

export const exec = call0
export const call = call1
