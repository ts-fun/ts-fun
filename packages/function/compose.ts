/**
 * Compose two function
 * @param f Function that takes `y` and returns `z`
 * @param g Function that takes `x` and returns `y`
 * @returns Function that takes `x` and returns `z = f(g(x))`
 */
export const compose = <X, Y, Z> (
  f: (y: Y) => Z,
  g: (x: X) => Y
) => (x: X) => f(g(x))

/**
 * Compose two function
 * @param f Function that takes `y` and returns `z`
 * @param g Function that takes `...xs` and returns `y`
 * @returns Function that takes `...xs` and returns `z = f(g(...xs))`
 */
export const composeXs = <Xs extends any[], Y, Z> (
  f: (y: Y) => Z,
  g: (...xs: Xs) => Y
) => (...xs: Xs) => f(g(...xs))

/**
 * Compose two function
 * @param f Function that takes `x` and returns `y`
 * @param g Function that takes `y` and returns `x`
 * @returns Function that takes `x` and returns `z = g(f(x))`
 */
export const composeRight = <X, Y, Z> (
  f: (x: X) => Y,
  g: (y: Y) => Z
) => compose(g, f)

/**
 * Compose two function
 * @param f Function that takes `...xs` and returns `y`
 * @param g Function that takes `y` and returns `z`
 * @returns Function that takes `...xs` and returns `z = g(f(...xs))`
 */
export const composeXsRight = <Xs extends any[], Y, Z> (
  f: (...xs: Xs) => Y,
  g: (y: Y) => Z
) => composeXs(g, f)

export {
  compose as composeFns,
  pipeline as composeFnsRight
} from 'ts-pipe-compose'
