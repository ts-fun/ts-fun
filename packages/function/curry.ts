export type Curriable<X, Xs extends any[], Y> = (x: X, ...xs: Xs) => Y
export type Curried<X, Xs extends any[], Y> = (...xs: Xs) => (x: X) => Y

/**
 * Postpone the first argument of a function
 * @param fn Function to curry
 * @returns Curried function
 */
export const curry =
  <X, Xs extends any[], Y>
    (fn: Curriable<X, Xs, Y>): Curried<X, Xs, Y> =>
      (...xs) => x => fn(x, ...xs)

export default curry
