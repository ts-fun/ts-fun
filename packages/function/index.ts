export * from './apply'
export * from './arity'
export * from './call'
export * from './compose'
export * from './curry'
export * from './partial'
export * from './pipe'
