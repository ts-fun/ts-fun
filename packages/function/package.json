{
  "name": "@tsfun/function",
  "version": "0.0.4",
  "description": "Function utilities",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://tsfun.github.io/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tsfun/tsfun.git"
  },
  "bugs": {
    "url": "https://github.com/tsfun/tsfun/issues"
  },
  "keywords": [
    "apply",
    "arity",
    "call",
    "compose",
    "curry",
    "partial",
    "pipe",
    "typescript"
  ],
  "sideEffects": false,
  "dependencies": {
    "ts-pipe-compose": "^0.1.11",
    "tslib": "^1.10.0",
    "@types/node": "^13.1.7"
  }
}
