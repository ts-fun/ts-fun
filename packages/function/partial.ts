/**
 * Partially apply a function
 * @param fn Function to apply
 * @param x First argument of `fn`
 * @returns Function that provides remaining arguments to `fn`
 */
export const partial = <X, Xs extends any[], Y> (
  fn: (x: X, ...xs: Xs) => Y,
  x: X
) => (...xs: Xs) => fn(x, ...xs)

/**
 * Partially apply a function
 * @param fn Function to apply
 * @param xs Array of all arguments except the first one
 * @returns Function that provides the first argument to `fn`
 */
export const partialTail = <X, Xs extends readonly any[], Y> (
  fn: (x: X, ...xs: Xs) => Y,
  xs: Xs
) => (x: X) => fn(x, ...xs)

/**
 * Partially apply a function
 * @param fn Function to apply
 * @param xs All arguments except the first one
 * @returns Function that provides the first argument to `fn`
 */
export const partialTailSpread = <X, Xs extends readonly any[], Y> (
  fn: (x: X, ...xs: Xs) => Y,
  ...xs: Xs
) => partialTail(fn, xs)
