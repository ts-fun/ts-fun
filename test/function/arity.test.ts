import { flip, flipXs, nAry, nAryRight } from '@tsfun/function'

describe('flip', () => {
  function setup () {
    const impl = (a: any, b: any) => [a, b]
    const fn = jest.fn(impl)
    const flipped = flip(fn)
    return { impl, fn, flipped }
  }

  function setupAndCall () {
    const { flipped, ...rest } = setup()
    const xs = ['left', 'right'] as const
    const y = flipped(...xs)
    return { ...rest, flipped, xs, y }
  }

  describe('when resulting function is not called', () => {
    it('does not call given function', () => {
      const { fn } = setup()
      expect(fn).not.toBeCalled()
    })
  })

  describe('when resulting function is called once', () => {
    it('calls given function once', () => {
      const { fn } = setupAndCall()
      expect(fn).toBeCalledTimes(1)
    })

    it('calls given function with swapped argument', () => {
      const { fn, xs: [left, right] } = setupAndCall()
      expect(fn).toBeCalledWith(right, left)
    })

    it('returns expected result', () => {
      const { impl, xs: [left, right], y } = setupAndCall()
      expect(y).toEqual(impl(right, left))
    })
  })
})

describe('flipXs', () => {
  function setup () {
    const impl = (...xs: any[]) => xs.join(', ')
    const fn = jest.fn(impl)
    const flipped = flipXs(fn)
    return { impl, fn, flipped }
  }

  function setupAndCall () {
    const { flipped, ...rest } = setup()
    const xs = ['a', 'b', 'c', 'd'] as const
    const y = flipped(...xs)
    return { ...rest, flipped, xs, y }
  }

  describe('when resulting function is not called', () => {
    it('does not call given function', () => {
      const { fn } = setup()
      expect(fn).not.toBeCalled()
    })
  })

  describe('when resulting function is called once', () => {
    it('calls given function once', () => {
      const { fn } = setupAndCall()
      expect(fn).toBeCalledTimes(1)
    })

    it('calls given function with swapped argument', () => {
      const { fn, xs: [left, right, ...rest] } = setupAndCall()
      expect(fn).toBeCalledWith(right, left, ...rest)
    })

    it('returns expected result', () => {
      const { impl, xs: [left, right, ...rest], y } = setupAndCall()
      expect(y).toEqual(impl(right, left, ...rest))
    })
  })
})

describe('nAry', () => {
  function setup () {
    const impl = (a: string, b: string) => `fn(${a}, ${b})`
    const fn = jest.fn(impl)
    const fnXs = nAry(fn)
    return { impl, fn, fnXs }
  }

  function setupAndCall () {
    const { fnXs, ...rest } = setup()
    const xs = ['a', 'b', 'c', 'd', 'e', 'f'] as const
    const y = fnXs(...xs)
    return { ...rest, fnXs, xs, y }
  }

  describe('when resulting function is not called', () => {
    it('does not call given function', () => {
      const { fn } = setup()
      expect(fn).not.toBeCalled()
    })
  })

  describe('when resulting function is called', () => {
    it('calls given function', () => {
      const { fn } = setupAndCall()
      expect(fn).toBeCalled()
    })

    it('calls given function multiple times', () => {
      const { fn } = setupAndCall()
      expect(fn.mock.calls).toMatchSnapshot()
    })

    it('returns expected result', () => {
      const { impl, xs, y } = setupAndCall()
      const [a, b, c, d, e, f] = xs
      expect(y).toBe(impl(impl(impl(impl(impl(a, b), c), d), e), f))
    })
  })
})

describe('nAryRight', () => {
  function setup () {
    const impl = (a: string, b: string) => `fn(${a}, ${b})`
    const fn = jest.fn(impl)
    const fnXs = nAryRight(fn)
    return { impl, fn, fnXs }
  }

  function setupAndCall () {
    const { fnXs, ...rest } = setup()
    const xs = ['a', 'b', 'c', 'd', 'e', 'f'] as const
    const y = fnXs(...xs)
    return { ...rest, fnXs, xs, y }
  }

  describe('when resulting function is not called', () => {
    it('does not call given function', () => {
      const { fn } = setup()
      expect(fn).not.toBeCalled()
    })
  })

  describe('when resulting function is called', () => {
    it('calls given function', () => {
      const { fn } = setupAndCall()
      expect(fn).toBeCalled()
    })

    it('calls given function multiple times', () => {
      const { fn } = setupAndCall()
      expect(fn.mock.calls).toMatchSnapshot()
    })

    it('returns expected result', () => {
      const { impl, xs, y } = setupAndCall()
      const [a, b, c, d, e, f] = xs
      expect(y).toBe(impl(impl(impl(impl(impl(f, e), d), c), b), a))
    })
  })
})
