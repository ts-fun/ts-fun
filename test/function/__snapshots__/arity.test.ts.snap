// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nAry when resulting function is called calls given function multiple times 1`] = `
Array [
  Array [
    "a",
    "b",
  ],
  Array [
    "fn(a, b)",
    "c",
  ],
  Array [
    "fn(fn(a, b), c)",
    "d",
  ],
  Array [
    "fn(fn(fn(a, b), c), d)",
    "e",
  ],
  Array [
    "fn(fn(fn(fn(a, b), c), d), e)",
    "f",
  ],
]
`;

exports[`nAryRight when resulting function is called calls given function multiple times 1`] = `
Array [
  Array [
    "f",
    "e",
  ],
  Array [
    "fn(f, e)",
    "d",
  ],
  Array [
    "fn(fn(f, e), d)",
    "c",
  ],
  Array [
    "fn(fn(fn(f, e), d), c)",
    "b",
  ],
  Array [
    "fn(fn(fn(fn(f, e), d), c), b)",
    "a",
  ],
]
`;
